staging/src/k8s.io/apimachinery/pkg/runtime/codec.go

```diff
@@ -43,10 +44,18 @@ func NewCodec(e Encoder, d Decoder) Codec {
 	return codec{e, d}
 }
 
+var bufPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
 // Encode is a convenience wrapper for encoding to a []byte from an Encoder
 func Encode(e Encoder, obj Object) ([]byte, error) {
-	// TODO: reuse buffer
-	buf := &bytes.Buffer{}
+	// allocates a new buffer or grabs a cached one
+	buf := bufPool.Get().(*bytes.Buffer)
```

Comment:
you are returning buf.Bytes() , that is a slice , and then returning the buffer to the pool.
I think that you have to copy them to a new slice before returning
```go
	bufferClone := make([]byte, buf.Len())
	copy(bufferClone, buf.Bytes())
        return bufferClone,nil
```

Reply:
Agree with the comment, but then this PR won't have any effect, right?

The whole point was to avoid allocating the buffer over and over and again - if we have to allocate that memory anyway, this improvement won't have any reasonable effect.

Reply:
Exactly! I ignored it. 

I added buffer copy but found after I did it, it had to allocate once memory when did copy action. 

It seems no improvement if I introduce sync.Pool and make buffer copy.